Add sending multiple CTE with periodic advertising chains

ppryga-nordic opened the pull request:
The PR enhances support for connectionless direction finding with possibility to transmit more than single PDU including CTE.
The multiple CTEs are transmitted with periodic advertising PDUs chain.

If the number of available PDUs in a periodic advertising chain is enough then CTE is just added to requested number of PDUs.
If the number of available PDUs in a chain is lower than number of requested CTEs, then the chain is extended with empty AUX_CHAIN_IND PDUs.

The PR is based on PR: https://github.com/zephyrproject-rtos/zephyr/pull/31875
It may be merged after the base PR is merged to main.

ppryga-nordic committed:
Bluetooth: controller: LLL: Add peek extra_data related with lll_adv_sync

Configuration of CTE for transmission is stored in extra_date member
of a lll_adv_sync object. PDUs in periodic advertising chain share the
same CTE configuration, so there is single instance of the extra_data.
To configure CTE transmission for every PDU in periodic advertising
chain a new function was introduced to peek extra_data instance without
peeking new PDU.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: df: Add const to antenna patterns pointer

Add const qualifier in radio API related with setting antenna
switch pattern.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add macro with minimum allowed adv. PDU size

To check if advertising PDU is empty we can compare its length to 1.
To avoid use of magic number, the commit provides a macro for that
purpose.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add conf max num of PDUs with CTE in per adv chain

Add possiblity to configure maximum number of PDUs with Constant Tone
Extension in a single periodic advertising chain.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Make per. adv. PDUs count related to max CTE num

Amount of memory allocated for advertising PDUs (including
periodic advertising) depends on two factors:
- maximum advertising data length
- maximum number of CTE in a periodic advertising chain.

Maximum advertising data length is divided by maximum size
of a single fragment (number of payload bytes that single
advertising PDU may hold) to get required number of fragments.

Actual number of PDUs allocated for advertising is maximum
of acutal number of advertising payload fragments and maximum
number of CTEs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: radio: Add back2back TX switch after PHYEND event

Transmission of Constant Tone Extension is done after END event
triggered by radio. To correctly switch from TX to TX (back 2 back
TX) after transmission of CTE PHYEND event must be used instead of
END event.

The commit provides required function that allows LLL to correctly
setup radio to do back to back switch of TX when CTE transmission
is enabled.

The radio_switch_complete_and_phy_end_b2b_tx cannot be implemented
in radio.c source file. It will fail build of unit tests targeted
for NRFBSIM simulator board. There is no RADIO_SHORTS_PHYEND_DISABLE_Msk
macro defined in bsim_hw_models module.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: Add CTE confing for per. adv. chain TX

Add configuration of CTE for periodic advertising chain transmission.
The commit provides configuration of radio in prepare stage and
in handle of Tx ISR. CTE is configured only for PDUs that
have cte_info field in extended advertising header.

During prepare of periodic advertising event there are updated
aux_ptr fields in extended advertising header in other PDUs from
a periodic advertising chain. aux_ptr offset value also depends
on CTE length. CTE configuration is always the same for every
PDU in periodic advertising chain.
CTE may be added to requested number of PDUs in periodic
advertising chain. Although it is possible that there are PDUs
that don't have CTE. PDUs that have CTE are alsways at the beginning
of a periodic advertising chain.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: Make API for per adv sync PDU change global

Some functions, provided to handle changing content of periodic
advertising PDUs, were defined as static.
Code responsible for handling direction finding updates periodic
advertising PDUs also. For that purposes, those functions were mede
globally accessible.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: ull_adv_sync_hdr_set_clear update PDU in place

Enabling or disabling transmission of Constant Tone Extension with
periodic advertising PDUs requires update of a PDU content.
CTE_INFO field of periodic advertising PDU must be filled with
appropriate data. That operation may be done for periodic
advertising PDU (or chain of PDUs) filled with other payload that
wasn't enqueued towards LLL. In that situation PDUs are updated
in place, to avoid increase of memory consumption.

The commit changes ull_adv_sync_pdu_set_clear function to
make possible update of the advertising PDU in place.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: Add sending of number of CTEs in per adv chain

Direction finding functionality allows to send a number of periodic
advertising PDUs in a chain that have CTE.
ll_df_set_cl_cte_tx_enable function was changed to update periodic
advertising chain to include cte_info field. If the chain is too short,
there is less PDUs in the chain than requested number of CTEs,
the function will add new empty PDUs to the chain end.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add release of chains when get latest adv pdu

Add release of chained PDUs by lll_adv_pdu_and_extra_data_-
latest_get function. It is requier to release unused
PDUs from a chain to avoid PDUs leakage.
It maight happen when chained PDUs were used by Direction
Finding, then lll_adv_pdu_and_extra_data_lates_get is used
instead of lll_adv_pdu_lates_get.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: Add handling of chains in disable CTE

Direction finding functionality allows to send a number of periodic
advertising PDUs in a chain that have CTE.
Disable sending CTE requiers additional steps while removing
cte_info from periodic advertising chains.
Removal of cte_info fields may be just delete of that filed
from extended advertising header. In case the PDUs are empty
PDUs created just to transport CTE. Those PDUs should be removed
from a periodic advertising chain.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Add tests for use of CTE in per adv chains

Bluetooth host may request to send more than one CTE in
connectionless mode. That is relized by periodic advertising
chained PDUs.
To check correctness of implementation of such functionality
new unit tests were added.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -98,6 +98,24 @@ static int init_reset(void);
+ *
+ * @param adv           Pointer to periodic advertising set.
+ * @param df_cfg        Pointer to direction finding configuration
+ * @param ter_idx[out]  Pointer used to return index of allocated or updated PDU.
```
Should be 
```
@param[out] ter_idx
```

ppryga replied:
Corrected.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -98,6 +98,24 @@ static int init_reset(void);
 static struct lll_df_adv_cfg *df_adv_cfg_acquire(void);
+
+/*
+ * @brief Function updates content of cte_info field in periodic advertising chain.
```
Description says `updates` but the name says `set`. I guess these should be aligned.

ppryga replied:
Aligned to the name of the function. Also I've moved the header to function definition, to align it with other functions.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +681,131 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+		return err;
+	}
+
+	if (extra_data) {
+		ull_adv_sync_extra_data_set_clear(extra_data_prev,
+						  extra_data,
+						  ULL_ADV_PDU_HDR_FIELD_CTE_INFO,
+						  0, &pdu_data);
+	}
```
If there is no `extra_data` do we need to clear the previous data?

ppryga replied:
Actually `extra_data_prev` is not needed here because is it not used. Every time CTE is set in a chain new configuration is set also. Besides that `ull_adv_sync_pdu_alloc` and `ull_adv_sync_pdu_set_clear` should be updated because they has unnecessary parameters.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +681,131 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+	}
+
+	err = ull_adv_sync_pdu_set_clear(lll_sync, pdu_prev, pdu,
+#if defined(CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX) && (CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1)
+					 (df_cfg->cte_count > 1) ?
+						       ULL_ADV_PDU_HDR_FIELD_CTE_INFO |
+							 ULL_ADV_PDU_HDR_FIELD_AUX_PTR :
+#endif /* CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX && CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1 */
+						       ULL_ADV_PDU_HDR_FIELD_CTE_INFO,
```
For readability sake, I think we could benefit from using a variable for this calculation.

ppryga replied:
Ah yes, I'll rework it a bit.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +681,131 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+	while (cte_index < df_cfg->cte_count) {
+		pdu_prev = pdu;
+		pdu = lll_adv_pdu_alloc_pdu_adv();
+		if (!pdu) {
+			return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+		}
```
I guess this also needs a todo similar to the TODO in lines 768-770

ppryga replied:
Yes, there is missing TODO about graceful error handling. The host is notified about problem here by return of `BT_HCI_ERR_MEM_CAPACITY_EXCEEDED`

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -32,6 +32,10 @@
 
 #define BYTES2US(bytes, phy) (((bytes)<<3)/BIT((phy&0x3)>>1))
 
+/* Advertisement channel minimum payload size */
+#define PDU_AC_PAYLOAD_SIZE_MIN 1
+/* Advertisement channel minimum payload size */
+#define PDU_AC_PAYLOAD_SIZE_MIN 1
```
Was one of these supposed to be a `MAX`?

ppryga replied:
Hm... I've got not idea why the macro is defined here twice. I'll remove one occurrence. `MAX` is not needed.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -81,6 +81,8 @@ static uint8_t mem_link_iq_report_quota_pdu;
 #if defined(CONFIG_BT_CTLR_DF_ADV_CTE_TX)
 static struct lll_df_adv_cfg lll_df_adv_cfg_pool[CONFIG_BT_CTLR_ADV_AUX_SET];
 static void *df_adv_cfg_free;
+static uint8_t remove_cte_info_from_per_adv_chain(struct ll_adv_set *adv,
+						  struct lll_df_adv_cfg *df_cfg, uint8_t *ter_idx);
```
Move `uint8_t *ter_idx` to the next line`

ppryga replied:
Although it looks wired, I suggest to leave it as is. This is a result of use of clang-format. It does not hurt but makes our lives easier without enable/disable automatic formatting in future.

Thalley replied:
From https://www.kernel.org/doc/html/latest/process/coding-style.html: 

> Descendants are always substantially shorter than the parent and are placed substantially to the right. A very commonly used style is to align descendants to a function open parenthesis.

Basically what this say, is that when breaking long lines, the lines should be shorts than the previous (when applicable of course).

ppryga replied:
Zephyr's checkpatch was aligned to that and reports warnings after exceeding 100 characters limit. .clang-format was also aligned to that and approved by community. So I assume that it is our limit now.

ppryga replied:
Moved to new line.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -81,6 +81,8 @@ static uint8_t mem_link_iq_report_quota_pdu;
 #if defined(CONFIG_BT_CTLR_DF_ADV_CTE_TX)
 static struct lll_df_adv_cfg lll_df_adv_cfg_pool[CONFIG_BT_CTLR_ADV_AUX_SET];
 static void *df_adv_cfg_free;
+static uint8_t remove_cte_info_from_per_adv_chain(struct ll_adv_set *adv,
```
Could we name `remove` to `rem` as this is already a pretty long function name?

ppryga replied:
Changed to `rem_cte_info_from_per_adv_chain`.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -808,4 +790,137 @@ static uint8_t set_cte_info_in_per_adv_chain(struct ll_adv_set *adv, struct lll_
+			return false;
+		}
+
+		const struct pdu_adv_ext_hdr *ext_hdr;
+		uint8_t size_rem = 0;
```
Not sure if this is an unwritten rule in the controller as well, but typically in the host (at least) we declare variables at top of scope.

ppryga replied:
Moved to the top of the scope

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -808,4 +790,137 @@ static uint8_t set_cte_info_in_per_adv_chain(struct ll_adv_set *adv, struct lll_
+		 * Add gracefull return or assert.
+		 */
+		return err;
+	}
+	return 0;
```
It often increases readability (slightly) if there is an empty line before `return` statements after a block like this

ppryga replied:
Added empty line before `return`

Thalley commented on tests/bluetooth/df/connectionless_cte_chains/src/common.c:
```diff
@@ -0,0 +1,472 @@
+ * The created chain is enqueued in provided advertising set. Number of requested PDUs includes
+ * head of a chain AUX_SYNC_IND.
+ * Each created PDU will hold payload data in following pattern:
+ * "test%d test%d test%d", where '%d' is substitured by PDU index.
```
`substitured` = > `substituted`

ppryga replied:
Corrected

Thalley commented on tests/bluetooth/df/connectionless_cte_chains/src/test_add_cte_to_chain.h:
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
2021

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -391,6 +391,66 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
 		return err;
 	}
 
+#if defined(CONFIG_BT_CTLR_ADV_PDU_LINK)
+	struct pdu_adv *ter_pdu = (void *)lll_sync->data.pdu[ter_idx];
```
Remember to remove this commit. I'll try to avoid approving the PR while it is still here.

ppryga replied:
Noted :-)

ppryga commented on subsys/bluetooth/controller/Kconfig.df:
```diff
@@ -146,6 +146,15 @@ config BT_CTLR_DF_INIT_ANT_SEL_GPIOS
+config BT_CTLR_DF_PER_ADV_CTE_NUM_MAX
+	int "Maximum number of transmitted PDUs with Constant Tone Extension in connectionelss mode"
+	depends on BT_CTLR_DF_ADV_CTE_TX
+	range 1 16
```
The range originates from BT Core spec 5.1, Vol 4 Part E section 7.8.82 HCI_LE_Set_Connectionless_IQ_Sampling_Enable. See at Max_Sampled_CTEs parameter. I've added this information to help section of the config.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -138,6 +138,13 @@ static inline struct pdu_adv *lll_adv_sync_data_peek(struct lll_adv_sync *lll,
+#if defined(CONFIG_BT_CTLR_ADV_EXT_PDU_EXTRA_DATA_MEMORY)
+static inline void *lll_adv_sync_extra_data_peek(struct lll_adv_sync *lll)
+{
+	return  lll->data.extra_data[lll->data.last];
```
Corrected, thanks.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -360,8 +349,7 @@ static void isr_tx(void *param)
 	radio_gpio_lna_setup();
-	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() +
-				 ADV_SYNC_PDU_B2B_AFS - 4 -
+	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() + ADV_SYNC_PDU_B2B_AFS - 4 + cte_len_us -
```
But current recommended line length is 100 characters. So it is in a limit.

Thalley replied:
Sure, it's nothing major. As far as I understood, the recommendation is still 80 character, but the limit was increased to 100 for some lines where readability can be improved. See https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=bdc48fa11e46f867ea4d75fa59ee87a7f48be144

ppryga replied:
Zephyr's checkpatch was aligned to that and reports warnings after exceeding 100 characters limit. .clang-format was also aligned to that and approved by community. So I assume that it is our limit now.

Thalley replied:
The limit has changed, but not the recommendation ;)

ppryga replied:
Oh.... let clang-format do its job and follow that unrecommended but allowed rule ;-)

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -42,19 +42,25 @@
 static int prepare_cb(struct lll_prepare_param *p);
 static void abort_cb(struct lll_prepare_param *prepare_param, void *param);
 static void isr_done(void *param);
+static inline void switch_radio_complete_and_b2b_tx(struct lll_adv_sync *lll, uint8_t phy_s);
+static inline void switch_radio_complete_and_phy_end_disable(struct lll_adv_sync *lll);
```
`inline` removed.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -42,19 +42,25 @@
-static void pdu_b2b_aux_ptr_update(struct pdu_adv *pdu, uint8_t phy,
-				   uint8_t flags, uint8_t chan_idx,
-				   uint32_t tifs);
+static inline void pdu_b2b_update(struct lll_adv_sync *lll, struct pdu_adv *pdu,
+				  uint32_t cte_len_us);
+static inline void pdu_b2b_aux_ptr_update(struct pdu_adv *pdu, uint8_t phy, uint8_t flags,
+					  uint8_t chan_idx, uint32_t offset_us,
+					  uint32_t cte_len_us);
```
Good point, thanks. `inline` removed.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -65,44 +67,60 @@ uint8_t lll_df_ant_num_get(void)
- * @Note Pay attention that first two antenna identifiers in a switch
- * pattern has special purpose. First one is used in guard period, second
- * in reference period. Actual switching is processed from third antenna.
- *
- * In case of AoA mode ant_num and ant_ids parameters are not used.
+ * @param lll_sync        Pointer to LLL sync. object associated with
+ *                        periodic advertising event.
+ * @param pdu             Pointer to PDU that will be transmitted.
+ * @param[out] cte_len_us Pointer to store actual CTE length in [us]
```
Yes, I share the option you have about where the function header comments belong. Although I was asked by @cvinayak to move comments like that from header file to source files: https://github.com/zephyrproject-rtos/zephyr/pull/29603#discussion_r530331691

Thalley replied:
Ah, then I'll leave it as is. He's the codeowner :)

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -207,7 +218,51 @@ void lll_df_conf_cte_rx_enable(uint8_t slot_duration, uint8_t ant_num, uint8_t *
+/* @brief Function initializes transmission of Constant Tone Extension.
+ *
+ * @param[in] type      Type of CTE: AoA, AoD 1us, AoD 2us
+ * @param[in] length    Duration of CTE in 8us units
```
Corrected.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -207,7 +218,51 @@ void lll_df_conf_cte_rx_enable(uint8_t slot_duration, uint8_t ant_num, uint8_t *
+ *
+ * @param[in] type      Type of CTE: AoA, AoD 1us, AoD 2us
+ * @param[in] length    Duration of CTE in 8us units
+ * @param[in] ant_num   Number of antennas in switch pattern
+ * @param[in] ant_ids   Antenna identifiers that create switch pattern.
```
Yes it is, `[in]` removed.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -856,10 +853,15 @@ uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync,
 	ter_com_hdr->adv_mode = ter_com_hdr_prev->adv_mode;
 	ter_hdr = (void *)ter_com_hdr->ext_hdr_adv_data;
-	ter_dptr = ter_hdr->data;
+	ter_dptr = &ter_hdr->data[0];
```
No, it is not. Change to former version.

ppryga replied:
Actually there was an issue in the function. It case of update of a PDU that hasn't got extended header, payload may be overwritten by code in line 865: `*(uint8_t *)ter_hdr = 0U;`. The extended advertising header may be stored in PDU only if there are some flags available (extended header length > 0).

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -207,7 +218,51 @@ void lll_df_conf_cte_rx_enable(uint8_t slot_duration, uint8_t ant_num, uint8_t *
+ * @param[in] ant_num   Number of antennas in switch pattern
+ * @param[in] ant_ids   Antenna identifiers that create switch pattern.
+ *
+ * @Note Pay attention that first two antenna identifiers in a switch
```
Both corrected.

ppryga replied:
Both corrected.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -207,7 +218,51 @@ void lll_df_conf_cte_rx_enable(uint8_t slot_duration, uint8_t ant_num, uint8_t *
+	if (type == BT_HCI_LE_AOA_CTE) {
+		radio_df_mode_set_aoa();
+		radio_df_cte_tx_aoa_set(length);
+	}
+#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX)
```
`radio_df_ant_switching_pin_sel_cfg` is dependent on both configs. The function has to be conditionally compiled because it depends on `ant_cfg` global variable that is also defined conditionally.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +681,131 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+			pdu_add_field_flags = 0;
+		} else {
+			++cte_index;
+			if (!pdu_next && cte_index < df_cfg->cte_count) {
```
I've added comment here, because it will be easier to understand this part of code.

ppryga commented:
@Thalley all comments were addressed.

I've to add a commit to clean up `ull_adv_sync_pdu_alloc` and `ull_adv_sync_pdu_set_clear` functions. I'll do that being back.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -359,6 +360,18 @@ void radio_switch_complete_and_phy_end_disable(void)
+			    RADIO_SHORTS_DISABLED_TXEN_Msk;
+#else /* !CONFIG_BT_CTLR_TIFS_HW */
+	NRF_RADIO->SHORTS = RADIO_SHORTS_READY_START_Msk | RADIO_SHORTS_PHYEND_DISABLE_Msk;
+	sw_switch(1, 1, phy_curr, flags_curr, phy_next, flags_next);
```
What does direction does the value `1` represent? 
I suggest making `#define`s that describe the semantic value

ppryga replied:
Changed, used macros instead of magic numbers W_SWITCH_PREV_TX, SW_SWITCH_NEXT_TX.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -65,44 +67,60 @@ uint8_t lll_df_ant_num_get(void)
- * In case of AoA mode ant_num and ant_ids parameters are not used.
+ * @param lll_sync        Pointer to LLL sync. object associated with
+ *                        periodic advertising event.
+ * @param pdu             Pointer to PDU that will be transmitted.
+ * @param[out] cte_len_us Pointer to store actual CTE length in [us]
  */
-void lll_df_conf_cte_tx_enable(uint8_t type, uint8_t length,
-			       uint8_t ant_num, uint8_t *ant_ids)
+void lll_df_cte_tx_enable(struct lll_adv_sync *lll_sync, const struct pdu_adv *pdu,
+			  uint32_t *cte_len_us)
```
What is the reasoning of providing a `uint32_t` out-pointer, rather than returning the value?

ppryga replied:
I thought about returning the cte_len_us by `return` keyword, but in my humble opinion use of pointer is more meaningful. You have the name of the pointer that suggests what is the purpose. If I had seen call to `lll_df_cte_tx_enable` I would expect it returns error code by 'return` statement.

Thalley replied:
> If I had seen call to lll_df_cte_tx_enable I would expect it returns error code by 'return` statement.

That is a fair point.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -402,17 +385,47 @@ static void pdu_b2b_aux_ptr_update(struct pdu_adv *pdu, uint8_t phy,
 }
 #endif /* CONFIG_BT_CTLR_ADV_SYNC_PDU_BACK2BACK */
+
+static void switch_radio_complete_and_b2b_tx(struct lll_adv_sync *lll, uint8_t phy_s)
```
`lll` could be const here

ppryga replied:
Added `const`

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -402,17 +385,47 @@ static void pdu_b2b_aux_ptr_update(struct pdu_adv *pdu, uint8_t phy,
+	}
+}
+
+static void switch_radio_complete_and_phy_end_disable(struct lll_adv_sync *lll)
```
`lll` could be const here

ppryga replied:
Added `const`

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -65,44 +67,60 @@ uint8_t lll_df_ant_num_get(void)
-
-		if (type == BT_HCI_LE_AOD_CTE_1US) {
-			radio_df_cte_tx_aod_2us_set(length);
+	const struct pdu_adv_ext_hdr *ext_hdr;
+	const struct lll_df_adv_cfg *df_cfg;
```
These could be declared in the scope of the following `if` statement.

ppryga replied:
Split and moved to `if` statements to be declared it in the scope where variables are used.

Thalley reviewed (changes requested):
A few minor comments, otherwise LGTM

ppryga commented:
Added changes requested by @Thalley.
Rebased on current main to clean number of commits introduced by this PR.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -207,7 +219,51 @@ void lll_df_conf_cte_rx_enable(uint8_t slot_duration, uint8_t ant_num, uint8_t *
+/* @brief Function performs common steps for initialization and reset
+ * of Direction Finding LLL module.
+ *
+ * @return	Zero in case of success, other value in case of failure.
```
Only use spaces for indentation if anything but a tab comes first
```suggestion
 * @return    Zero in case of success, other value in case of failure.
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -207,7 +219,51 @@ void lll_df_conf_cte_rx_enable(uint8_t slot_duration, uint8_t ant_num, uint8_t *
+/* @brief Function initializes transmission of Constant Tone Extension.
+ *
+ * @param type      Type of CTE: AoA, AoD 1us, AoD 2 us
+ * @param length    Duration of CTE in 8us units
```
```suggestion
 * @param length    Duration of CTE in 8 us units
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +649,286 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+	 */
+	err = ull_adv_sync_pdu_alloc(adv, ULL_ADV_PDU_HDR_FIELD_CTE_INFO, 0, NULL, &pdu_prev, &pdu,
+				     NULL, &extra_data, ter_idx);
+	if (err) {
+		return err;
+	}
```
`err` is an `int` but this function returns an `uint8_t`. Missing either an explicit cast, or change the return type of this function.

ppryga replied:
Changed type of `err` to `uint8_t`. It stores error codes from other function calls that return `uint8_t`.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +649,286 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+
+		err = ull_adv_sync_pdu_set_clear(lll_sync, pdu_prev, pdu, pdu_add_field_flags, 0,
+						 &pdu_data);
+		if (err != 0) {
```
I generally like `if (err != 0)` better than `if (err)`, but you are being slightly inconsistent with this. Please update for consistency (preferably in this way, but either way is alright).

ppryga replied:
Changed to `if (err != 0)` here and in other `if` statements in the function.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +649,286 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+		if ((pdu->adv_ext_ind.ext_hdr_len - size_rem) != PDU_AC_EXT_HEADER_SIZE_MIN) {
+			return false;
+		}
+	}
+	return true;
```
```suggestion
	}
	
	return true;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -138,6 +138,13 @@ static inline struct pdu_adv *lll_adv_sync_data_peek(struct lll_adv_sync *lll,
 }
+
+#if defined(CONFIG_BT_CTLR_ADV_EXT_PDU_EXTRA_DATA_MEMORY)
+static inline void *lll_adv_sync_extra_data_peek(struct lll_adv_sync *lll)
+{
+	return lll->data.extra_data[lll->data.last];
+}
```
It would be rather unusual in controller code to have static non-inline function in header file. Is it related with this: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#the-inline-disease

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -352,8 +338,7 @@ static void isr_tx(void *param)
 
 	radio_gpio_lna_setup();
-	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() +
-				 ADV_SYNC_PDU_B2B_AFS - 4 -
+	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() + ADV_SYNC_PDU_B2B_AFS - 4 + cte_len_us -
 				 radio_tx_chain_delay_get(lll->phy_s, 0) -
```
This is 2 times `2 us` JITTER related with active clock accuracy. It is `4 us` because TX clock may be `-2 us` and receiver clock may be `+2 us`.

Thalley replied:
I think @cvinayak has changed a few in the controller in one of his PRs already based on a similar question from me. 

Since it was there before your commit, I won't let it block my approval, but it is probably a good idea to, at one point, change all these occurrences with a macro.

ppryga replied:
In my humble opinion it should be changed. I'll create a separate PR that adds such change through out the controller code.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_internal.h:
```diff
@@ -0,0 +1,16 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SW_SWITCH_PREV_RX 0
+#define SW_SWITCH_NEXT_RX 0
+#define SW_SWITCH_PREV_TX 1
+#define SW_SWITCH_NEXT_TX 1
+#define SW_SWITCH_PREV_PHY_1M 0
+#define SW_SWITCH_PREV_FLAGS_DONTCARE 0
+#define SW_SWITCH_NEXT_FLAGS_DONTCARE 0
```
Yes I agree. I've changed it according to suggestion.

ppryga commented:
@Thalley Two comments were not addresses. The one with `static inline` and the one related with `2 us` jitter. The latter requires separate commit and general change in controller code to remove jitter related magic numbers.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -689,4 +649,287 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+	 */
+	err = ull_adv_sync_pdu_alloc(adv, ULL_ADV_PDU_HDR_FIELD_CTE_INFO, 0, NULL, &pdu_prev, &pdu,
+				     NULL, &extra_data, ter_idx);
+	if (err != 0) {
+		return err;
+	}
```
Since `ull_adv_sync_pdu_alloc` returns an `BT_HCI_ERR_` code, we could change this check to 
`if (err != BT_HCI_ERR_SUCCESS)`, but that seems to not be done a lot in the controller (or at all), so perhaps it's best to be consistent.

ppryga replied:
I've changed it to `if (err != BT_HCI_ERR_SUCCESS)`. Lets stay consistent about return types and macros used for that.

ppryga commented:
Rebased on laters main.
Fixed two issues:
- PDU memory was released twice to `pdu_free` fifo and `mem_pdu`. Caused by function: `rem_cte_info_from_per_adv_chain` when CTE only PDU is found and end of chain is released.
- assignment of wrong address to `ter_dptr_prev` in `ull_adv_sync_pdu_set_clear`. Caused evaluation of wrong size of ACAD and increase of PDU size. Eventually PDU size reached limit and `BT_HCI_ERR_PACKET_TOO_LONG` was returned.

ppryga commented:
@cvinayak @andrzej-kaczmarek @Thalley  could you take a look?

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -32,6 +32,8 @@
 #define BYTES2US(bytes, phy) (((bytes)<<3)/BIT((phy&0x3)>>1))
 
+/* Advertisement channel minimum payload size */
+#define PDU_AC_PAYLOAD_SIZE_MIN 1
```
Was any code supposed to be updated to use this?

ppryga replied:
yes it is used in one of later commits: https://github.com/zephyrproject-rtos/zephyr/blob/dcfde9b3acdd4e3dd9b4e398601ca8a364b88a15/subsys/bluetooth/controller/ll_sw/ull_df.c#L792

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_internal.h:
```diff
@@ -0,0 +1,13 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SW_SWITCH_RX 0
+#define SW_SWITCH_TX 1
+#define SW_SWITCH_PHY_1M 0
+#define SW_SWITCH_FLAGS_DONTCARE 0
```
Looking more into how these are used, it looks like the RX/TX values are used for the `dir_curr` and `dir_next` exclusively. 
Similarly, the `PHY_1M` is used exclusively for the `phy_next` and the `FLAGS_DONTCARE`. 

Based on the naming, this isn't clear. I suggest to, either in this or future PR, either
1) Move these into enums and use the enum types in  `sw_switch` 
2) Add an infix in the macro to make their use more clear, e.g. `SW_SWITCH_DIR_RX` and `SW_SWITCH_DIR_TX`

ppryga replied:
I'd like to avid further factorization of this code in the PR. The goal of the PR is a bit different. I suggest to create a github issue for code improvements and store there all of those comments.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_internal.h:
```diff
@@ -0,0 +1,13 @@
+#define SW_SWITCH_PHY_1M 0
+#define SW_SWITCH_FLAGS_DONTCARE 0
+
+void sw_switch(uint8_t dir_curr, uint8_t dir_next, uint8_t phy_curr, uint8_t flags_curr,
+	       uint8_t phy_next, uint8_t flags_next);
```
The order of the arguments is a bit odd as they go:
dir, dir, phy, flags, phy, flags

It would make more sense if it was
dir, phy, flags, dir, phy, flags
or
dir, dir, phy, phy, flags, flags

Or possible using a struct
```
struct sw_switch_param {
    uint8_t dir;
    uint8_t phy;
    uint8_t flags;
};

void sw_switch(struct sw_switch_param curr, struct sw_switch_param next);
```

ppryga replied:
1. See this comment: https://github.com/zephyrproject-rtos/zephyr/pull/36585/commits/6c7bba67f157170e3642cfe9644af09b5c477164#r677372071
2. I've observed that is it kind of a pattern in controller to go for long parameters list instead of structures. @cvinayak can you give more feedback about your choices in this regard?

Thalley replied:
I don't see any comment in your reference :s 

And I agree that it has been a typical pattern in the controller code.

ppryga replied:
Sorry, I've taken link to comment from commit instead of conversation page.
This is correct one: https://github.com/zephyrproject-rtos/zephyr/pull/36585#discussion_r677372071

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -65,44 +67,61 @@ uint8_t lll_df_ant_num_get(void)
+			df_cfg = lll_adv_sync_extra_data_peek(lll_sync);
+			LL_ASSERT(df_cfg);
+
+			df_cte_tx_configure(df_cfg->cte_type, df_cfg->cte_length,
+					    df_cfg->ant_sw_len, df_cfg->ant_ids);
```
You use enough fields of the `df_cfg` struct that it might make sense to actually use `const struct lll_df_adv_cfg *df_cfg` as the argument to the function, unless there are scenarios where we need to call this function without the access to such a struct?

ppryga replied:
Changed to struct instead of list of parameters.
Interesting thing, I did some measurements. Flash usage after the change decreased by 16B while number of instructions related with the `df_cte_tx_configure` was larger bu 3 instructions (hope I didn't missed any instruction while counting).

Thalley replied:
That's curious; unless you updated the function besides the parameters, it should have resulted to the same.

ppryga replied:
Body of the function `df_cte_tx_configure` has changed due to changed arguments list. 
Take also into consideration that compiler may inline the body of the function. That may save additional bytes.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -665,4 +635,311 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+
+	/* Go through existing chain and remove CTE info. */
+	while (pdu_chained != NULL) {
+		if (pdu_ext_adv_is_empty_without_cte(pdu_chained) == true) {
```
```suggestion
		if (pdu_ext_adv_is_empty_without_cte(pdu_chained)) {
```

ppryga replied:
Code was refactored in a similar way as here: https://github.com/zephyrproject-rtos/zephyr/pull/36585/commits/1f6c24c1c6ed5efc3e7f7c8e653f8ca0f8cf3d08#r675458125

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -1094,6 +1099,10 @@ uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync,
 	/* No TargetA in AUX_SYNC_IND */
 	/* No AdvA in AUX_SYNC_IND */
 
+	if (ter_com_hdr->ext_hdr_len) {
```
Updated to explicit comparison. I didn't touch other places like https://github.com/zephyrproject-rtos/zephyr/blob/4e485c756525c148d2354810e8d84e443bd6217f/subsys/bluetooth/controller/ll_sw/ull_adv_sync.c#L1090
because there is similar function in ull_adv_aux.c: https://github.com/zephyrproject-rtos/zephyr/blob/4e485c756525c148d2354810e8d84e443bd6217f/subsys/bluetooth/controller/ll_sw/ull_adv_aux.c#L442 that has similar pattern.
In both cases the code should not pass MISRA check. It would be better to do it in separate PR where is clear information why the changes is done.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -665,4 +648,154 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+
+	*first_pdu = pdu;
+
+#if defined(CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX) && (CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1)
```
The `CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX` should be guarded by `CONFIG_BT_CTLR_ADV_PDU_LINK`. I'll fix that. It can't be removed because the function uses `lll_adv_pdu_linked_next_get` function that is compiled conditionally.

ppryga replied:
I've extracted appropriated part of code and created separate function.

Thalley replied:
> It can't be removed because the function uses lll_adv_pdu_linked_next_get function that is compiled conditionally.

As long as the function has a header, it doesn't matter if it is compiled or not, as that would be optimized/handled by the linker.

ppryga replied:
Yes, in this case linker will complain about unresolved `lll_adv_pdu_linked_next_get` reference.

Thalley replied:
Then you must also be guarding `lll_adv_pdu_linked_next_get`, right? 

It's fine to keep as is, so no need to spend more time on that.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -665,4 +648,154 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+	ull_adv_sync_extra_data_set_clear(NULL, extra_data, ULL_ADV_PDU_HDR_FIELD_CTE_INFO, 0,
+					  df_cfg);
+
+#if defined(CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX) && (CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1)
+	if (df_cfg->cte_count > 1) {
```
See https://github.com/zephyrproject-rtos/zephyr/pull/36585/commits/1f6c24c1c6ed5efc3e7f7c8e653f8ca0f8cf3d08#r677593811

Thalley replied:
I'm not sure what to see in that reference - At least for me it just points to the file in the commit :D

ppryga replied:
This is correct comment reference: https://github.com/zephyrproject-rtos/zephyr/pull/36585#discussion_r677593811

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -665,4 +648,154 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+
+	/* Update PDUs in existing chain. Add cte_info to extended advertising header. */
+	while (pdu_prev != NULL) {
+		if (new_chain == true) {
```
Removed explitict comparison. I did the same for line 729 due to checkpatch complains.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -665,4 +648,154 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+	}
+
+	/* If there is missing only one CTE do not add aux_ptr to PDU */
+	if (df_cfg->cte_count - cte_index >= 2) {
```
No, `cte_index` is local variable. It is initalized to `1` and incremented only if `if (cte_index >= df_cfg->cte_count)`.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -665,4 +634,363 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+	ull_adv_sync_extra_data_set_clear(NULL, extra_data, ULL_ADV_PDU_HDR_FIELD_CTE_INFO, 0,
+					  df_cfg);
+
+#if (CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1)
+	if (df_cfg->cte_count > 1) {
+		pdu_add_field_flags =
+			ULL_ADV_PDU_HDR_FIELD_CTE_INFO | ULL_ADV_PDU_HDR_FIELD_AUX_PTR;
+	} else
+#endif /* CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1 */
```
```suggestion
	if (CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1 && 
	    df_cfg->cte_count > 1) {
		pdu_add_field_flags =
			ULL_ADV_PDU_HDR_FIELD_CTE_INFO | ULL_ADV_PDU_HDR_FIELD_AUX_PTR;
	} else
```

ppryga replied:
This is not possible. `BT_CTLR_DF_PER_ADV_CTE_NUM_MAX`  dependends on `BT_CTLR_ADV_PDU_LINK`. If one wanted to use single CTE and didn't enable chaining PUDs that code would not build.

Thalley reviewed (approved):
Only a minor suggestion, otherwise LGTM

cvinayak reviewed (changes requested):
Very minor change requests, otherwise looks good to me. (Have not tested on target).

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -885,7 +884,7 @@ uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync,
 {
 	struct pdu_adv_com_ext_adv *ter_com_hdr, *ter_com_hdr_prev;
-	struct pdu_adv_ext_hdr *ter_hdr, ter_hdr_prev;
+	struct pdu_adv_ext_hdr ter_hdr = { 0 }, ter_hdr_prev = { 0 };
```
iirc, structs can only be initialized when declared, so if we remove the initialization here, we'd need to a `memset` later or set the individual fields to 0.

ppryga replied:
Change to `memset` increases FLASH size by 16 bytes in comparison to current version. Unfortunately I haven't got J-Trace to measure it during execution. I'm leaving it as it is implemented to avoid increase of FLASH size.

cvinayak replied:
The header is 1 bytes, use `memset` and compiler will optimize it to `*((uint8_t *)ter_hdr_prev) = 0U;`

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -81,6 +81,8 @@ static uint8_t mem_link_iq_report_quota_pdu;
 static struct lll_df_adv_cfg lll_df_adv_cfg_pool[CONFIG_BT_CTLR_ADV_AUX_SET];
 static void *df_adv_cfg_free;
+static uint8_t rem_cte_info(struct ll_adv_set *adv, struct lll_df_adv_cfg *df_cfg, uint8_t *ter_idx,
+			    struct pdu_adv **first_pdu);
```
Changed according to suggestion

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -98,6 +100,9 @@ static int init_reset(void);
 static struct lll_df_adv_cfg *df_adv_cfg_acquire(void);
+
+static uint8_t set_cte_info(struct ll_adv_set *adv, struct lll_df_adv_cfg *df_cfg, uint8_t *ter_idx,
+			    struct pdu_adv **first_pdu);
```
Changed according to suggestion

ppryga-nordic committed:
Bluetooth: controller: remove warning about defined bvt unused function

Remove compilation warning about adv_sync_pdu_ad_data_set function
that is defined but not used.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -665,4 +634,363 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
+ */
+static uint8_t set_cte_info_in_per_adv_chain(struct lll_adv_sync *lll_sync,
+					     struct pdu_adv *pdu_prev, struct pdu_adv *pdu,
+					     uint8_t cte_count, struct pdu_cte_info *cte_info)
```
Changed according to suggestion

ppryga commented:
Applied changes requested by @cvinayak.
Added commit that removes compilation warning about unused function: `adv_sync_pdu_ad_data_set`.

Code was tested by `direction_finding_connectionless_tx` sample and Bluetooth sniffer.
